lib/spack/spack/install_test.py

```diff
@@ -52,6 +93,638 @@ def get_test_stage_dir():
     )
 
 
+def cache_extra_test_sources(pkg: "spack.package_base.PackageBase", srcs: ListOrStringType):
+    """Copy relative source paths to the corresponding install test subdir
+
+    This routine is intended as an optional install test setup helper for
+    grabbing source files/directories during the installation process and
+    copying them to the installation test subdirectory for subsequent use
+    during install testing.
+
+    Args:
+        pkg: package being tested
+        srcs: relative path for file(s) and or subdirectory(ies) located in
+            the staged source path that are to be copied to the corresponding
+            location(s) under the install testing directory.
+    """
+    paths = [srcs] if isinstance(srcs, str) else srcs
+
+    for path in paths:
+        src_path = os.path.join(pkg.stage.source_path, path)
+        dest_path = os.path.join(install_test_root(pkg), path)
+        if os.path.isdir(src_path):
+            fs.install_tree(src_path, dest_path)
+        else:
+            fs.mkdirp(os.path.dirname(dest_path))
+            fs.copy(src_path, dest_path)
+
+
+def check_outputs(expected: Union[list, set, str], actual: str):
+    """Ensure the expected outputs are contained in the actual outputs.
+
+    Args:
+        expected: expected raw output string(s)
+        actual: actual output string
+
+    Raises:
+        RuntimeError: the expected output is not found in the actual output
+    """
+    expected = expected if isinstance(expected, (list, set)) else [expected]
+    for check in expected:
+        if not re.search(check, actual):
+            raise RuntimeError("Expected '{0}' in output '{1}'".format(check, actual))
```

Comment:
Would it be better to collect all the failures and report them afterwards?

Reply:
IIRC we weren't doing that before but I like that thinking.

Reply:
Done.